yield_plugin/src/yield_plugin.cpp

```diff
@@ -0,0 +1,252 @@
+    : wm_(wm), config_(config), plugin_discovery_publisher_(plugin_discovery_publisher)
+  {
+    plugin_discovery_msg_.name = "YieldPlugin";
+    plugin_discovery_msg_.versionId = "v1.0";
+    plugin_discovery_msg_.available = true;
+    plugin_discovery_msg_.activated = false;
+    plugin_discovery_msg_.type = cav_msgs::Plugin::TACTICAL;
+    plugin_discovery_msg_.capability = "tactical_plan/plan_trajectory";
+  }
+
+  bool YieldPlugin::onSpin()
+  {
+    plugin_discovery_publisher_(plugin_discovery_msg_);
+    return true;
+  }
+
+  bool YieldPlugin::plan_trajectory_cb(cav_srvs::PlanTrajectoryRequest& req,
+                                              cav_srvs::PlanTrajectoryResponse& resp)
+  {
+    if (req.initial_trajectory_plan.trajectory_points.size() < 2){
+      throw std::invalid_argument("Empty Trajectory received");
+    }
+    cav_msgs::TrajectoryPlan original_trajectory = req.initial_trajectory_plan;
+    cav_msgs::TrajectoryPlan trajectory;
+    trajectory.header.frame_id = "map";
+    trajectory.header.stamp = original_trajectory.header.stamp;// should it be now or the original plan's time? 
+    trajectory.trajectory_id = original_trajectory.trajectory_id;
+
+    trajectory = update_traj_for_object(original_trajectory, req.vehicle_state.longitudinal_vel); // Compute the trajectory
+    trajectory.initial_longitudinal_velocity = trajectory.initial_longitudinal_velocity;//std::max(req.vehicle_state.longitudinal_vel, config_.minimum_speed);//????  find better value
+
+    resp.trajectory_plan = trajectory;
+  }
+
+  
+  cav_msgs::TrajectoryPlan YieldPlugin::update_traj_for_object(const cav_msgs::TrajectoryPlan& original_tp, double current_speed_) 
+  {
+        
+    cav_msgs::TrajectoryPlan update_tpp_vector;
+    geometry_msgs::Twist current_velocity;
+    current_velocity.linear.x = current_speed_;
+
+    std::vector<cav_msgs::RoadwayObstacle> rwol = wm_->getRoadwayObjects();
+    cav_msgs::RoadwayObstacleList rwol2;
+    rwol2.roadway_obstacles = rwol;
+    host_vehicle_size.x = config_.vehicle_length;
+    host_vehicle_size.y = config_.vehicle_width;
+    host_vehicle_size.z = config_.vehicle_height; 
+    // std::cout << "host_vehicle_size" << host_vehicle_size.x << std::endl;
+    std::vector<cav_msgs::RoadwayObstacle> rwol_collision = carma_wm::collision_detection::WorldCollisionDetection(rwol2, original_tp, host_vehicle_size, current_velocity, config_.collision_horizon);
+    std::cout << "rwol" << rwol.size() << std::endl;
+
+    // correct the input types
+    if(rwol_collision.size() > 0)
+    {
+      ROS_WARN_STREAM("Collision!");
+
+      // Distance from the original trajectory point to the lead vehicle/object
+      double dist_x = rwol_collision[0].object.pose.pose.position.x - original_tp.trajectory_points[0].x;
+      double dist_y = rwol_collision[0].object.pose.pose.position.y - original_tp.trajectory_points[0].y;
+      double x_lead = sqrt(dist_x*dist_x + dist_y*dist_y);
```

Comment:
Is this going to be accurate on curves? Pose is in map frame right? Should we be concerned if the road curvature might cause this distance to be calculated as shorter than the path along the lane curvature?

Reply:
I agree that it is not accurate curvey roads, but since this distance would be smaller than the actual value, I believe it beneficial from the safety perspective.

Reply:
I agree that the distance calculation is not accurate, but it is helpful in this case because the distance is smaller than the actual value, and makes the vehicle decelerate without getting too close to the obstacle.